### DISCUSSION TITLE
Define endpoints independently of the bind address

### DISCUSSION THREAD

**Discussion by ondrejnovakcvut** [FreeOpcUa/opcua-asyncio#674] [Ideas]
Hello. When I wanted to deploy freeopcua server to Docker, I encountered a problem that has been described several times (for example #232 ). The problem is that now, the endpoint description returned to the client by the discovery are modified to match clients IP range ([here](https://github.com/FreeOpcUa/opcua-asyncio/blob/fc09e423c7f3e49783b0788ad0933fd836175c15/asyncua/server/internal_server.py#L206)). In Docker, where clients are "hidden" behind a NAT, the server doesn't know client's real IP addresses so the rewrite creates non-existent endpoints in virtual docker network (e.g., 172.16.x.x).

I think if there was 
a) a way to define server bind independently of endpoint IP addresses,
b) a way to manually create endpoints (by specifying IP and Security Profiles),
c) disable the automatic IP rewrite,
it would be possible to overcome this.

For example, the programmer could set the bind address to "0.0.0.0" to listen inside the Docker, then the programmer could define multiple endpoints for two IP addresses "1.2.3.4" and "5.6.7.8" with anon and sign policies. The client would then get four endpoints, two for each IP address. Only problem I can see is the client would have to select the relevant endpoint that it can connect to as not all IP address may be accessible by the client.

I think I would be able to implement this, but I would like to discuss the idea and the (possible) API first.

**Comment by oroulet** [comment]
Ok. We need to fix that. Cannot be hard.

**Comment by oroulet** [comment]
OK. I am not sure I understand exactly the issue and I am quite short on time, so correct me if I am wrong
One sure thing is that we can add an argument to disable IP change in get_endpoint(). This is anyway an improvement. Do you make a PR for that?
But does that solve your issue?
Also there is nothing preventing you to call Server.iserver.add-endpoint() and add your endpoint manually

**Reply by oroulet** [reply]
get_endpoints(return_raw_ip=True) for example

**Reply by ondrejnovakcvut** [reply]
Disabling the IP change solves it only partially. 

Default endpoint is `opc.tcp://0.0.0.0:4840/freeopcua/server/`, that would bind the server to `0.0.0.0` ([here](https://github.com/FreeOpcUa/opcua-asyncio/blob/fc09e423c7f3e49783b0788ad0933fd836175c15/asyncua/server/server.py#L390)), which is fine. The problem is when the server starts, function ` _setup_server_nodes` ([here](https://github.com/FreeOpcUa/opcua-asyncio/blob/fc09e423c7f3e49783b0788ad0933fd836175c15/asyncua/server/server.py#L311)) is called. The function creates default endpoints with IP `0.0.0.0`, which doesn't make a sense in this context. When the IP rewrite is disabled, the client would receive non-existent endpoints. Also, it may be a bit confusing to set the IP address that the server is bound to by setting the default endpoint, that won't even exist.

I'm aware of the possibility to call the `add_endpoint()` function directly. Nevertheless, this function is fairly complex, as it requires `EndpointDescription` object. I think a simplifying API that would create this object for you would be nice. This is not really necessary to solve my problem (so it could be a separate PR), but if I'm making some modifications to the endpoint API, I should probably keep these (possible) modifications in mind.

To make it nice and easy, I think of possibility to: 
- set the bind address explicitly and disable the automatic IP change (kwarg in the `Server` constructor)
- define endpoints by `server.add_endpoint(url: str, policies: List[str], security_policy: ua.SecurityPolicy, mode: ua.MessageSecurityMode)` (I'd just be reusing `server._add_endpoints()`) in a separate PR

**Reply by oroulet** [reply]
OK try . seems to make sense to me
